pkg/sql/show_fingerprints.go

```diff
@@ -0,0 +1,169 @@
+// (`::string::bytes`) and is an obvious area for improvement in the next
+// version.
+//
+// AS OF SYSTEM TIME is optional, but when set, uses the table definition and
+// data as of that the specified time.
+func (p *planner) ShowFingerprints(
+	ctx context.Context, n *parser.ShowFingerprints,
+) (planNode, error) {
+	ts := p.session.execCfg.Clock.Now()
+	if n.AsOf.Expr != nil {
+		evalCtx := p.session.evalCtx()
+		var err error
+		ts, err = EvalAsOfTimestamp(&evalCtx, n.AsOf, ts)
+		if err != nil {
+			return nil, err
+		}
+	}
+	p.txn.SetFixedTimestamp(ts)
+	p.avoidCachedDescriptors = true
+
+	tn, err := n.Table.NormalizeWithDatabaseName(p.session.Database)
+	if err != nil {
+		return nil, err
+	}
+	tableDesc, err := mustGetTableDesc(ctx, p.txn, p.getVirtualTabler(), tn)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := p.CheckPrivilege(tableDesc, privilege.SELECT); err != nil {
+		return nil, err
+	}
+	return &showFingerprintsNode{
+		p:         p,
+		n:         n,
+		tn:        tn,
+		tableDesc: tableDesc,
+		indexes:   tableDesc.AllNonDropIndexes(),
+	}, nil
+}
+
+type showFingerprintsNode struct {
+	p *planner
+	n *parser.ShowFingerprints
+
+	tn        *parser.TableName
+	tableDesc *sqlbase.TableDescriptor
+	indexes   []sqlbase.IndexDescriptor
+
+	rowIdx int
+	// values stores the current row, updated by Next().
+	values []parser.Datum
+}
+
+var showFingerprintsColumns = sqlbase.ResultColumns{
+	{
+		Name: "Index",
+		Typ:  parser.TypeString,
+	},
+	{
+		Name: "Fingerprint",
+		Typ:  parser.TypeString,
+	},
+}
+
+func (n *showFingerprintsNode) Start(ctx context.Context) error { return nil }
+func (n *showFingerprintsNode) Values() parser.Datums           { return n.values }
+func (n *showFingerprintsNode) Close(_ context.Context)         {}
+func (n *showFingerprintsNode) Next(ctx context.Context) (bool, error) {
+	if n.rowIdx >= len(n.indexes) {
+		return false, nil
+	}
+	index := n.indexes[n.rowIdx]
+
+	columns := make([]string, 0, len(n.tableDesc.Columns))
+	addColumn := func(col sqlbase.ColumnDescriptor) {
+		// TODO(dan): This is known to be a flawed way to fingerprint. Any datum
+		// with the same string representation is fingerprinted the same, even
+		// if they're different types.
+		switch col.Type.Kind {
+		case sqlbase.ColumnType_BYTES:
+			columns = append(columns, col.Name)
+		default:
+			columns = append(columns, fmt.Sprintf("%s::string::bytes", col.Name))
+		}
+	}
+
+	if index.ID == n.tableDesc.PrimaryIndex.ID {
+		for _, col := range n.tableDesc.Columns {
+			addColumn(col)
+		}
+	} else {
+		colsByID := make(map[sqlbase.ColumnID]sqlbase.ColumnDescriptor)
+		for _, col := range n.tableDesc.Columns {
+			colsByID[col.ID] = col
+		}
+		colIDs := append(append(index.ColumnIDs, index.ExtraColumnIDs...), index.StoreColumnIDs...)
+		for _, colID := range colIDs {
+			col := colsByID[colID]
+			addColumn(col)
+		}
+	}
+
+	sql := fmt.Sprintf(`SELECT
+	  TO_HEX(XOR_AGG(FNV64(%s))) AS fingerprint
```

Comment:
Even though *we're* not persisting these yet, I'd love to put a marker in the fingerprint, just so if we change this later we have some easy-ish way to figure out what we're looking at.

Maybe `concat('fx.', xor_agg(...)::string)`?